### ISSUE TITLE
RVV buggy when using IREE generated binaries

### ISSUE THREAD

**Issue by alhussein-jamil** [gem5/gem5#2145]
**Describe the bug**

When simulating a RISC-V binary generated by IREE (which executes a simple PyTorch matrix transpose model), gem5 produces incorrect results. The task is to transpose an 8x8 matrix of f32 values initialized to 1.0. The expected output is an 8x8 matrix of 1.0s. However, the simulation yields a matrix where only the top half contains 1.0s, and the bottom half contains 0.0s.

This suggests a potential issue in gem5's RISC-V Vector Extension (RVV) implementation, possibly related to the handling of the vector length multiplier (`lmul`) or specific vector instructions like `vslidedown`/`vslideup`. The simulation uses `VLEN=128` and the IREE compilation specifies `--riscv-v-fixed-length-vector-lmul-max=8`. Given the input data type is f32 (`SEW=32`), `VLEN=128` allows processing 4 elements per vector register (`VLMAX = VLEN / SEW = 128 / 32 = 4`). The incorrect result (only the first 4 rows seem correct) might indicate that operations involving `lmul > 1` (using multiple vector registers for a single logical vector operation) are not handled correctly.

The workflow involves:
1.  Defining a simple `transpose` operation in PyTorch.
2.  Exporting the model to MLIR using `torch.export`.
3.  Compiling the MLIR to a RISC-V executable module (`.vmfb`) using the IREE compiler, targeting RVV.
4.  Running this module using IREE's RISC-V runtime (`iree-run-module`) within a gem5 simulation (SE mode).

**PyTorch Model:**
```python
import torch

class Transpose(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x.transpose(0, 1)

# Example usage (not run in gem5, just for context)
# model = Transpose()
# input_tensor = torch.ones(8, 8, dtype=torch.float32) # Input tensor of ones
# output_tensor = model(input_tensor)
# print(output_tensor) # Should print an 8x8 tensor of ones
```

**IREE Compilation Command:**
```bash
# Assuming ${mlir_file} points to the exported MLIR representation
# Assuming ${binary_name} is transpose.vmfb
iree-compile "${mlir_file}" \
  --iree-hal-target-backends=llvm-cpu \
  --iree-llvmcpu-target-triple=riscv64-unknown-elf \
  --iree-llvmcpu-target-abi=lp64 \
  --iree-llvmcpu-target-cpu-features=+m,+a,+f,+zvl128b,+v \
  --riscv-v-fixed-length-vector-lmul-max=8 \
  -o "${binary_name}"
```

**gem5 Modifications**
None.

**To Reproduce**

**Prerequisites:**
*   A working gem5 build for RISC-V.
*   A RISC-V toolchain (e.g., the one used for IREE compilation).
*   IREE compiled for RISC-V host tools (`iree-compile`) and target runtime (`iree-run-module`).
*   The compiled model artifact (`transpose.vmfb`) generated using the IREE command above from a suitable MLIR input.
*   The RISC-V `iree-run-module` binary.

**Steps:**

1.  **Compile gem5:**
    ```bash
    # From the gem5 root directory
    scons build/RISCV/gem5.opt -j<N> # Replace <N> with number of cores
    ```

2.  **Prepare Input Files:**
    *   Place the compiled IREE module `transpose.vmfb` in the simulation run directory.
    *   Place the `iree-run-module` binary (compiled for RISC-V `lp64d` ABI) in the simulation run directory or adjust the path in the script.
    *   Ensure the RISC-V toolchain sysroot path is correct in the script's environment variables.

3.  **Create gem5 Configuration Script (`reproduce.py`):**
    Save the following script (adjust paths as necessary, especially `riscv_toolchain_root` and the `binary` path):

    ```python
    import m5
    from m5.objects import *
    import numpy as np
    import os

    # Create a simple system with a RISC-V O3 CPU
    system = System()
    system.clk_domain = SrcClockDomain(clock="1GHz", voltage_domain=VoltageDomain())
    system.mem_mode = "timing"
    system.mem_ranges = [AddrRange("512MiB")]

    # CPU configuration
    system.cpu = RiscvO3CPU()
    # Ensure the RiscvISA is created before accessing its properties
    system.cpu.isa = [RiscvISA()]
    # Enable RVV
    system.cpu.isa[0].enable_rvv = True
    system.cpu.isa[0].vlen = 128  # Vector register length in bits
    system.cpu.isa[0].elen = 64   # Max Element length in bits (Note: actual SEW is 32 for f32)

    # Memory system
    system.membus = SystemXBar()
    system.cpu.icache_port = system.membus.cpu_side_ports
    system.cpu.dcache_port = system.membus.cpu_side_ports
    system.cpu.createInterruptController()

    # Simpler memory controller for SE mode consistency
    system.mem_ctrl = MemCtrl()
    system.mem_ctrl.dram = DDR3_1600_8x8()
    system.mem_ctrl.dram.range = system.mem_ranges[0]
    system.mem_ctrl.port = system.membus.mem_side_ports

    system.system_port = system.membus.cpu_side_ports # Connect system port for functional access

    # Process configuration
    process = Process()
    # Ensure the binary path is correct
    binary_path = "iree-run-module" # Assumes it's in PATH or PWD
    if not os.path.exists(binary_path):
         # Try relative path often used in build dirs or specify full path
         alt_path = os.path.join(os.getcwd(), "iree-run-module")
         if os.path.exists(alt_path):
             binary_path = alt_path
         else:
             # FAIL LOUDLY - USER MUST PROVIDE CORRECT PATH
             raise FileNotFoundError(f"Cannot find iree-run-module binary. Searched: {binary_path}, {alt_path}")

    # Check if the workload binary exists and is compatible
    system.workload = SEWorkload.init_compatible(binary_path)

    # Command to execute
    process.cmd = [
        binary_path,
        "--module=transpose.vmfb",
        "--device=local-sync",
        "--function=main",
        "--input=8x8xf32=1.0", # Input matrix of ones
        "--output=@output.npy", # Save output to file
    ]

    # --- IMPORTANT: Adjust Paths Below ---
    # Set this to the root of your RISC-V GCC/Clang toolchain installation
    # Example: /opt/riscv/toolchain/clang/linux/RISCV/
    riscv_toolchain_root = "/path/to/your/riscv/toolchain/root/"
    if not os.path.isdir(riscv_toolchain_root):
        raise NotADirectoryError(f"RISCV toolchain root not found: {riscv_toolchain_root}")

    sysroot_path = os.path.join(riscv_toolchain_root, "sysroot")
    if not os.path.isdir(sysroot_path):
        raise NotADirectoryError(f"Sysroot not found: {sysroot_path}")

    # Assuming lp64d ABI library path - adjust if using a different ABI (e.g., lp64)
    lib_path = os.path.join(sysroot_path, "lib64", "lp64d")
    if not os.path.isdir(lib_path):
         # Fallback for lp64 if lp64d doesn't exist
         lib_path_lp64 = os.path.join(sysroot_path, "lib")
         if os.path.isdir(lib_path_lp64):
             lib_path = lib_path_lp64
         else:
             raise NotADirectoryError(f"RISCV library path not found: Checked {lib_path} and {lib_path_lp64}")

    libm_path = os.path.join(lib_path, "libm.so.6")
    if not os.path.isfile(libm_path):
         # Try finding any libm.so*
         import glob
         found_libm = glob.glob(os.path.join(lib_path, "libm.so*"))
         if found_libm:
             libm_path = found_libm[0] # Use the first one found
         else:
             raise FileNotFoundError(f"libm.so.6 or similar not found in {lib_path}")
    # --- End Path Adjustment ---


    # Environment variables needed by IREE runtime
    process.env = [
        "IREE_BINARY_ROOT=./", # Where to find modules relative to (if needed)
        f"RISC_V_SYSROOT={sysroot_path}",
        f"LD_LIBRARY_PATH={lib_path}",
        "IREE_DEFAULT_CPU_DRIVER=dylib", # Tells IREE how to load functions
        f"IREE_CPU_DYLIB_LIBM_PATH={libm_path}",
    ]

    # Set workload and create threads
    system.cpu.workload = process
    system.cpu.createThreads()

    # Setup root and instantiate
    root = Root(full_system=False, system=system)
    m5.instantiate()

    # Run simulation
    print("Beginning simulation!")
    exit_event = m5.simulate()
    exit_cause = exit_event.getCause()
    print(f"Exiting @ tick {m5.curTick()} because {exit_cause}")

    # Check results if simulation finished normally
    if "exiting with last active thread context" in exit_cause and os.path.exists("output.npy"):
        try:
            output = np.load("output.npy")
            print("Simulation output (output.npy):")
            print(output)
        except Exception as e:
            print(f"Could not load or print output.npy: {e}")
    else:
        print("Simulation did not exit normally or output.npy not found.")

    ```

4.  **Execute the simulation:**
    ```bash
    ./build/RISCV/gem5.opt reproduce.py
    ```

5.  **Observe Output:** Check the terminal output and the contents of the generated `output.npy` file.

**Terminal Output (Relevant Sections)**
```shell
gem5 Simulator System.  https://www.gem5.org
# ... (gem5 version/build info) ...
Global frequency set at 1000000000000 ticks per second
# ... (warnings about DRAM size, symtab collisions - likely unrelated) ...
src/arch/riscv/isa.cc:XXX: info: RVV enabled, VLEN = 128 bits, ELEN = 64 bits # Actual line might differ slightly based on version
# ... (syscall warnings - likely unrelated) ...
Beginning simulation!
src/sim/simulate.cc:XXX: info: Entering event queue @ 0.  Starting simulation...
# ... (stack increase info, more syscall warnings) ...
EXEC @main # Output from iree-run-module indicating function execution
Exiting @ tick 16214114000 because exiting with last active thread context # Tick count will vary
Simulation output (output.npy):
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0.]]
```

**Expected behavior**
The simulation should complete successfully, and the `output.npy` file (and the printed NumPy array at the end of the simulation) should contain an 8x8 matrix filled entirely with `1.0`.

```python
# Expected output content
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]]
```

      
*   **Verification with QEMU:** The *same* `iree-run-module` binary and `transpose.vmfb` module execute correctly under QEMU (version 8.1.2) using settings comparable to the gem5 configuration. This further suggests the issue might be specific to gem5's RVV implementation. The QEMU command used for verification is:
    ```bash
    # Ensure RISCV_TOOLCHAIN_ROOT is set correctly
    # Ensure binary_name points to transpose.vmfb
    # inputs_arg should contain the input specification, e.g., "--input=8x8xf32=1.0 --output=@output_qemu.npy"
    # env should contain necessary LD_LIBRARY_PATH etc. as used in the gem5 script

    qemu-riscv64 \
        -cpu rv64,zve64d=true,vlen=128,elen=64,vext_spec=v1.0 \
        -L ${RISCV_TOOLCHAIN_ROOT}/sysroot/ \
        ./iree-run-module \
        --module=${binary_name} \
        --function=main \
        ${inputs_arg}
        # Note: The --parameter_mode='preload' might be specific to certain IREE versions/setups
        # and the --output flag is needed to check the result.
    ```
    Running this produces the expected output matrix of all ones.

  

**Additional information**
*   The issue seems tied to vector operations spanning more elements than fit in a single vector register (`VLEN/SEW = 128/32 = 4`), suggesting `lmul > 1` or related addressing/masking/slide operations might be incorrectly simulated.
*   The IREE compilation explicitly allows `lmul` up to 8. It would be helpful to know which specific vector instructions are generated by IREE for the `transpose` operation in this configuration.
*   The simulation uses `RiscvO3CPU`. It might be worth testing with `MinorCPU` or `AtomicSimpleCPU` (if RVV is supported) to see if the issue persists, which could help isolate whether it's in the core functional implementation or related to the O3 pipeline specifics.
```

**Comment by erin-le** [comment]
@alhussein-jamil Hello, thanks for opening this issue. Based on your description of the bug, it's possible that this issue could be fixed by https://github.com/gem5/gem5/pull/1955, which has been merged into the `develop` branch, or by https://github.com/gem5/gem5/pull/1712, which is still open. Could you try using PR 1712 and the develop branch and see if they work for you?

**Comment by alhussein-jamil** [comment]
@erin-le Hello, we were trying to use some other framework rather than IREE, I have already been using develop,  I tried #1712 patch, it does not seem to fix the problem. 
I am still getting exactly the same result.

**Comment by saul44203** [comment]
Hi @alhussein-jamil, could you check if #2307 fixes this? If not, I would appreciate if you could upload a static binary, that way it's easier for me to reproduce it.

**Comment by erin-le** [comment]
@alhussein-jamil: Hello, just a reminder about this. It'd be great if we could get https://github.com/gem5/gem5/pull/2307 into the gem5 v25.0 release. The staging branch for the release has already been created, but bug fixes are still being accepted onto the staging branch. The staging branch will be merged into stable on June 12th at the earliest, so it'd be helpful if you could take a look at it soon. Thank you!
